File: Source/AlphaTab/Rendering/Glyphs/AlternateEndingsGlyph.cs
using AlphaTab.Collections;
using AlphaTab.Model;
using AlphaTab.Platform;
using AlphaTab.Platform.Model;

namespace AlphaTab.Rendering.Glyphs
{
    internal class AlternateEndingsGlyph : EffectGlyph
    {
        private const float Padding = 3;

        private readonly FastList<int> _endings;
        private string _endingsString;

        public AlternateEndingsGlyph(float x, float y, byte alternateEndings)
            : base(x, y)
        {
            _endings = new FastList<int>();
            for (var i = 0; i < MasterBar.MaxAlternateEndings; i++)
            {
                if ((alternateEndings & (0x01 << i)) != 0)
                {
                    _endings.Add(i);
                }
            }
        }

        public override void DoLayout()
        {
            base.DoLayout();
            Height = Renderer.Resources.WordsFont.Size + (Padding * Scale + 2);
            var endingsStrings = new StringBuilder();
            for (int i = 0, j = _endings.Count; i < j; i++)
            {
                endingsStrings.Append(_endings[i] + 1);
                endingsStrings.Append(". ");
            }

            _endingsString = endingsStrings.ToString();
        }

        public override void Paint(float cx, float cy, ICanvas canvas)
        {
            base.Paint(cx, cy, canvas);

            var baseline = canvas.TextBaseline;
            canvas.TextBaseline = TextBaseline.Top;

            if (_endings.Count > 0)
            {
                var res = Renderer.Resources;
                canvas.Font = res.WordsFont;
                canvas.MoveTo(cx + X, cy + Y + Height);
                canvas.LineTo(cx + X, cy + Y);
                canvas.LineTo(cx + X + Width, cy + Y);
                canvas.Stroke();

                canvas.FillText(_endingsString, cx + X + Padding * Scale, cy + Y * Scale);
            }

            canvas.TextBaseline = baseline;
        }
    }
}

File: src/rendering/glyphs/AlternateEndingsGlyph.ts
import { type ICanvas, TextBaseline } from '@src/platform/ICanvas';
import { EffectGlyph } from '@src/rendering/glyphs/EffectGlyph';
import type { RenderingResources } from '@src/RenderingResources';
import { ModelUtils } from '@src/model/ModelUtils';

export class AlternateEndingsGlyph extends EffectGlyph {
    private static readonly Padding: number = 3;
    private _endings: number[];
    private _endingsString: string = '';
    private _openLine: boolean;
    private _closeLine: boolean;

    public constructor(x: number, y: number, alternateEndings: number, openLine: boolean, closeLine: boolean) {
        super(x, y);
        this._endings = ModelUtils.getAlternateEndingsList(alternateEndings);
        this._openLine = openLine;
        this._closeLine = closeLine;
    }

    public override doLayout(): void {
        super.doLayout();
        this.height = this.renderer.resources.wordsFont.size + (AlternateEndingsGlyph.Padding + 2);
        let endingsStrings: string = '';
        for (let i: number = 0, j: number = this._endings.length; i < j; i++) {
            endingsStrings += this._endings[i] + 1;
            endingsStrings += '. ';
        }
        this._endingsString = endingsStrings;
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        const width = this._closeLine ? this.width - 4 : this.width;

        cx = (cx | 0) + 0.5;
        cy = (cy | 0) + 0.5;

        if (this._openLine) {
            canvas.moveTo(cx + this.x, cy + this.y + this.height);
            canvas.lineTo(cx + this.x, cy + this.y);
        } else {
            canvas.moveTo(cx + this.x, cy + this.y);
        }

        canvas.lineTo(cx + this.x + width, cy + this.y);

        if (this._closeLine) {
            canvas.lineTo(cx + this.x + width, cy + this.y + this.height);
        }

        canvas.stroke();

        if (this._openLine) {
            const baseline: TextBaseline = canvas.textBaseline;
            canvas.textBaseline = TextBaseline.Top;
            const res: RenderingResources = this.renderer.resources;
            canvas.font = res.wordsFont;
            canvas.fillText(this._endingsString, cx + this.x + AlternateEndingsGlyph.Padding, cy + this.y);
            canvas.textBaseline = baseline;
        }
    }
}
